File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/library/items/stored/specs/GivenAnUnfoundCollectionOFServiceFiles/WhenCollectingTheAssociatedServiceFiles.java
package com.lasthopesoftware.bluewater.client.library.items.stored.specs.GivenAnUnfoundCollectionOFServiceFiles;

import com.annimon.stream.Stream;
import com.lasthopesoftware.bluewater.client.library.items.IItem;
import com.lasthopesoftware.bluewater.client.library.items.media.files.ServiceFile;
import com.lasthopesoftware.bluewater.client.library.items.media.files.access.IFileProvider;
import com.lasthopesoftware.bluewater.client.library.items.media.files.access.parameters.FileListParameters;
import com.lasthopesoftware.bluewater.client.library.items.stored.IStoredItemAccess;
import com.lasthopesoftware.bluewater.client.library.items.stored.StoredItem;
import com.lasthopesoftware.bluewater.client.library.items.stored.StoredItemServiceFileCollector;
import com.lasthopesoftware.messenger.promises.Promise;

import org.assertj.core.api.Condition;
import org.junit.BeforeClass;
import org.junit.Test;

import java.io.FileNotFoundException;
import java.util.Arrays;
import java.util.Collection;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Random;
import java.util.concurrent.CountDownLatch;

import static com.annimon.stream.Stream.concat;
import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class WhenCollectingTheAssociatedServiceFiles {

	private static Collection<ServiceFile> collectedFiles;
	private static List<ServiceFile> firstItemExpectedFiles = givenARandomCollectionOfFiles();
	private static List<ServiceFile> secondItemExpectedFiles = givenARandomCollectionOfFiles();
	private static List<ServiceFile> thirdItemExpectedFiles = givenARandomCollectionOfFiles();
	private static HashMap<IItem, Boolean> syncToggledItems = new HashMap<>();

	@BeforeClass
	public static void before() throws InterruptedException {

		final IStoredItemAccess storedItemAccess =
			new IStoredItemAccess() {
				@Override
				public void toggleSync(IItem item, boolean enable) {
					syncToggledItems.put(item, enable);
				}

				@Override
				public Promise<Boolean> isItemMarkedForSync(IItem item) {
					return null;
				}

				@Override
				public Promise<Collection<StoredItem>> promiseStoredItems() {
					return new Promise<>(Arrays.asList(
						new StoredItem(1, 1, StoredItem.ItemType.ITEM),
						new StoredItem(1, 2, StoredItem.ItemType.ITEM),
						new StoredItem(1, 3, StoredItem.ItemType.ITEM)));
				}
			};

		final IFileProvider fileProvider = mock(IFileProvider.class);
		when(fileProvider.promiseFiles(FileListParameters.Options.None, "Browse/Files", "ID=1"))
			.thenAnswer(e -> new Promise<>(firstItemExpectedFiles));
		when(fileProvider.promiseFiles(FileListParameters.Options.None, "Browse/Files", "ID=2"))
			.thenAnswer(e -> new Promise<>(new FileNotFoundException()));
		when(fileProvider.promiseFiles(FileListParameters.Options.None, "Browse/Files", "ID=3"))
			.thenAnswer(e -> new Promise<>(thirdItemExpectedFiles));

		final StoredItemServiceFileCollector serviceFileCollector = new StoredItemServiceFileCollector(
			storedItemAccess,
			fileProvider);

		final CountDownLatch countDownLatch = new CountDownLatch(1);
		serviceFileCollector
			.promiseServiceFilesToSync()
			.then(files -> {
				collectedFiles = files;
				countDownLatch.countDown();
				return null;
			})
			.excuse(e -> {
				countDownLatch.countDown();
				return null;
			});

		countDownLatch.await();
	}

	@Test
	public void thenOnlyTheFoundServiceFilesAreReturned() {
		assertThat(collectedFiles).hasSameElementsAs(new HashSet<>(concat(Stream.of(firstItemExpectedFiles), Stream.of(thirdItemExpectedFiles)).toList()));
	}

	@Test
	public void thenTheFileThatWasNotFoundHadSyncToggledOff() {
		assertThat(syncToggledItems).hasEntrySatisfying(new Condition<IItem>() {

			@Override
			public boolean matches(IItem value) {
				return value.getKey() == 2;
			}
		}, new Condition<Boolean>() {
			@Override
			public boolean matches(Boolean value) {
				return !value;
			}
		});
	}

	private static List<ServiceFile> givenARandomCollectionOfFiles() {
		final Random random = new Random();
		final int floor = random.nextInt(10000);
		final int ceiling = random.nextInt(10000 - floor) + floor;
		return Stream.range(floor, ceiling).map(ServiceFile::new).toList();
	}
}

File: projectBlueWater/src/test/java/com/lasthopesoftware/bluewater/client/stored/library/items/GivenAnUnfoundCollectionOfServiceFiles/WhenCollectingTheAssociatedServiceFiles.kt
package com.lasthopesoftware.bluewater.client.stored.library.items.GivenAnUnfoundCollectionOfServiceFiles

import com.lasthopesoftware.bluewater.client.browsing.files.ServiceFile
import com.lasthopesoftware.bluewater.client.browsing.files.access.ProvideLibraryFiles
import com.lasthopesoftware.bluewater.client.browsing.files.access.parameters.FileListParameters
import com.lasthopesoftware.bluewater.client.browsing.items.ItemId
import com.lasthopesoftware.bluewater.client.browsing.items.KeyedIdentifier
import com.lasthopesoftware.bluewater.client.browsing.library.repository.LibraryId
import com.lasthopesoftware.bluewater.client.stored.library.items.AccessStoredItems
import com.lasthopesoftware.bluewater.client.stored.library.items.FakeStoredItemAccess
import com.lasthopesoftware.bluewater.client.stored.library.items.StoredItem
import com.lasthopesoftware.bluewater.client.stored.library.items.StoredItemServiceFileCollector
import com.lasthopesoftware.bluewater.shared.promises.extensions.toExpiringFuture
import com.lasthopesoftware.promises.extensions.toPromise
import com.namehillsoftware.handoff.promises.Promise
import io.mockk.every
import io.mockk.mockk
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Condition
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test
import java.io.FileNotFoundException
import kotlin.random.Random.Default.nextInt

class WhenCollectingTheAssociatedServiceFiles {

	private val firstItemExpectedFiles by lazy { givenARandomCollectionOfFiles() }
	private val thirdItemExpectedFiles by lazy { givenARandomCollectionOfFiles() }
	private val syncToggledItems = HashMap<KeyedIdentifier, Boolean>()
	private val serviceFileCollector by lazy {
		val storedItemAccess: AccessStoredItems = object : FakeStoredItemAccess(
			StoredItem(4, 1, StoredItem.ItemType.ITEM),
			StoredItem(4, 2, StoredItem.ItemType.ITEM),
			StoredItem(4, 3, StoredItem.ItemType.ITEM)
		) {
			override fun toggleSync(libraryId: LibraryId, itemId: KeyedIdentifier, enable: Boolean): Promise<Unit> {
				syncToggledItems[itemId] = enable
				return super.toggleSync(libraryId, itemId, enable)
			}
		}

		val fileListParameters = FileListParameters
		val fileProvider = mockk<ProvideLibraryFiles>()
		every {
			fileProvider.promiseFiles(
				LibraryId(4),
				FileListParameters.Options.None,
				*fileListParameters.getFileListParameters(ItemId(1))
			)
		} returns firstItemExpectedFiles.toPromise()

		every {
			fileProvider.promiseFiles(
				LibraryId(4),
				FileListParameters.Options.None,
				*fileListParameters.getFileListParameters(ItemId(2))
			)
		} returns Promise(FileNotFoundException())

		every {
			fileProvider.promiseFiles(
				LibraryId(4),
				FileListParameters.Options.None,
				*fileListParameters.getFileListParameters(ItemId(3))
			)
		} returns thirdItemExpectedFiles.toPromise()

		StoredItemServiceFileCollector(
			storedItemAccess,
			fileProvider,
			fileListParameters
		)
	}

	private lateinit var collectedFiles: Collection<ServiceFile>

	@BeforeAll
	fun act() {
		collectedFiles = serviceFileCollector
			.promiseServiceFilesToSync(LibraryId(4))
			.toExpiringFuture().get()!!
	}

	@Test
	fun `then only the found service files are returned`() {
		assertThat(collectedFiles).hasSameElementsAs(firstItemExpectedFiles.plus(thirdItemExpectedFiles).toHashSet())
	}

	@Test
	fun `then the file that was not found had sync toggled off`() {
		assertThat(syncToggledItems).hasEntrySatisfying(object : Condition<KeyedIdentifier>() {
			override fun matches(value: KeyedIdentifier): Boolean = value.id == 2
		}, object : Condition<Boolean>() {
			override fun matches(value: Boolean): Boolean = !value
		})
	}

	private fun givenARandomCollectionOfFiles(): List<ServiceFile> {
		val floor = nextInt(10000)
		val ceiling = nextInt(10000 - floor) + floor
		return IntRange(floor, ceiling).map { ServiceFile(it) }.toList()
	}
}
